cron.py

```diff
@@ -757,6 +763,106 @@ def delete_old_monitoring():
     LOG.d("delete monitoring records older than %s, nb row %s", max_time, nb_row)
 
 
+def _hibp_check(api_key, queue):
+    """
+    Uses a single API key to check the queue as fast as possible.
+
+    This function to be ran simultaneously (multiple _hibp_check functions with different keys on the same queue) to make maximum use of multiple API keys.
+    """
+    while True:
+        try:
+            alias_id = queue.get_nowait()
+        except Empty:
+            return
+
+        alias = Alias.get(alias_id)
+        LOG.d("Checking HIBP for %s", alias)
+
+        request_headers = {
+            "user-agent": "SimpleLogin",
+            "hibp-api-key": api_key,
+        }
+        r = requests.get(
+            f"https://haveibeenpwned.com/api/v3/breachedaccount/{urllib.parse.quote(alias.email)}",
+            headers=request_headers,
+        )
+
+        if r.status_code == 200:
+            # Breaches found
+            alias.hibp_breaches = [Hibp.get_by(id=entry["Name"]) for entry in r.json()]
+        elif r.status_code == 404:
+            # No breaches found
+            alias.hibp_breaches = []
+        else:
+            LOG.error(
+                "An error occured while checking alias %s: %s - %s",
+                alias,
+                r.status_code,
+                r.text,
+            )
+            return
+
+        db.session.add(alias)
+        db.session.commit()
+
+        LOG.d("Updated breaches info for %s", alias)
+
+        sleep(1.5)
+
+
+def check_hibp():
+    """
+    Check all aliases on the HIBP (Have I Been Pwned) API
+    """
+    LOG.d("Checking HIBP API for aliases in breaches")
+
+    if len(HIBP_API_KEYS) == 0:
+        LOG.exception("No HIBP API keys")
+        return
+
+    LOG.d("Updating list of known breaches")
+    r = requests.get("https://haveibeenpwned.com/api/v3/breaches")
+    for entry in r.json():
+        Hibp.get_or_create(id=entry["Name"])
+
+    db.session.commit()
+    LOG.d("Updated list of known breaches")
+
+    LOG.d("Preparing list of aliases to check")
+    queue = multiprocessing.Queue()
```

Comment:
I wonder if we can split the alias to handle into `len(HIBP_API_KEYS)` chunks and launch a process for handling each chunk? I feel this way is a bit more simple to understand but this is probably because I'm not familiar with using Queue ...

Reply:
It is possible.

The benefit of my current method is though that if for some reason some web request ends up stalling (think: network hiccup and it ends up getting a timeout), it won't cause any slowdown as each of the API keys will grab an entry from the queue as soon as it is ready to do another request. If you split it in 4 beforehand and one web request would time out, it would be slower at the end. Not really a big difference, but splitting it up before starting makes the code more complex and introduces an odd edge cases with network timeouts.

Reply:
Oh I see, that makes sense 👌.

I'm wondering if asyncio can be a good fit for our usecase here? Your previous implementation is actually almost ready to  convert into a asyncio code :).

Reply:
Oh, perhaps. I just have never used asyncio before, so I really wouldn't know where to get started asyncio-ifying this

Reply:
Well, turns out it is actually super simple. Done in 40d0dee.

Reply:
Wow I love the new implementation!